### PR TITLE
Vanilla: Call onError without argument if there is no response

### PR DESCRIPTION
**Pull request by minad**
See https://github.com/haskell-servant/servant/issues/607

### CODE REVIEW

**Comment by minad** [comment]
ping @soenkehahn

**Comment by phadej** [comment]
- it would be good to check that Content-Type is`application/json`
- Also it would be great to `JSON.parse` inside `try` / `catch` block

**Comment by minad** [comment]
Ok, I am pretty sure this can be improved. My reasoning is that if JSON is served, it should be valid - this PR only treats the case of no response.

But I am still not sure about what is desired here:

1. Should onError called without an argument or with some kind of error object? If we want an error object, which format, e.g. `{ status: ..., msg: ... }`?
2. Should onError be called if JSON parsing fails or should there be an exception?

**Comment by phadej** [comment]
My concerns are that in not `2xx` case the response might be (and in case of say 500 exception, isn't) a valid JSON. So maybe I'd just `onError(new Error(xhr))`

**Comment by minad** [comment]
I updated the patch

File: src/Servant/JS/Vanilla.hs [hunk 1]
```diff
@@ -29,28 +29,28 @@ generateVanillaJS = generateVanillaJSWith defCommonGeneratorOptions
...
+ <> "      if (xhr.status === 204 || xhr.status === 205) {\n"
+ <> "        " <> onSuccess <> "();\n"
+ <> "      } else if (xhr.status >= 200 && xhr.status < 300 && xhr.responseType === 'json') {\n"
+ <> "        try { res = JSON.parse(xhr.responseText); } catch (e) { " <> onError <> "(e); }\n"
+ <> "        if (res) " <> onSuccess <> "(res);\n"
  <> "      } else {\n"
- <> "        var value = JSON.parse(xhr.responseText);\n"
- <> "        onError(value);\n"
+ <> "        try { res = JSON.parse(xhr.responseText); } catch (e) { " <> onError <> "(e); }\n"
+ <> "        if (res) " <> onSuccess <> "(res);\n"
```
Comment (phadej) [inline src/Servant/JS/Vanilla.hs]: `onSuccess`, it should be `onError`?

**Comment by minad** [comment]
ouch yes

**Review by phadej — APPROVED** [review]
LGTM, I'll try it and merge then. Thanks.

**Pull request merged**